### PR TITLE
Migrate Python CDK to Poetry

### PR DESCRIPTION
**Pull request by aaronsteers**
## What this does

This migrates the Airbyte Python CDK to [Poetry](https://python-poetry.org/), which is the modern standard for Python dev-time dependency resolution and virtual env management.

## Background

When starting an new connector, I immediately ran into a version conflict between the library's required version of the `jsonschema` library and the version that we support in the CDK, with the library requiring a newer version than our `jsonschema` constraints dictate.

Without Poetry, it is very difficult to debug dependencies and figure out how/if we can bump this particular `jsonschema` dependency. With poetry, it is very simple.

Since this is not he first time I ran into a similar issue, and since we are in Hackathon week, I decided to kick off this process and start the migration effort from `setup.py` to `Poetry` and `pyproject.toml`.

### CODE REVIEW

**Comment by aaronsteers** [comment]
This is ready for review (tests are passing), but there may be other integration points I'm not thinking of as of yet.

File: airbyte-cdk/python/pyproject.toml [hunk 1]
```diff
@@ -1,8 +1,91 @@
...
+avro = {version = ">=1.11.2,<1.12.0", optional = true}
+fastavro = {version = ">=1.8.0,<1.9.0", optional = true}
+pyarrow = {version = "12.0.1", optional = true}
+
+# File-based Extras (Unstructured Docs):
+markdown = {version = "^3.5.1", optional = true}
+pdf2image = {version = "1.16.3", optional = true}
+pdfminer-six = {version = "20221105", optional = true}
+pytesseract = {version = "0.3.10", optional = true}
+unstructured = {version = "0.10.27", extras = ["file-based"]} # Unstructured version can't be bumped higher due to transitive dependencies we can't provide:
```
Comment (flash1293) [inline airbyte-cdk/python/pyproject.toml:63]: why are extras called out in the dependency for these two, but specified in `tool.poetry.extras` for the others?

File: airbyte-cdk/python/README.md [hunk 2]
```diff
@@ -99,6 +106,7 @@ pip install -e .
 ```
 
 Then, navigate to the CDK and install it in editable mode:
+
 ```bash
 cd ../../../airbyte-cdk/python
 pip install -e .
```
Comment (flash1293) [inline airbyte-cdk/python/README.md:112]: Will this still work? It's a pretty important workflow (at least I'm using it daily), so we should make sure it works with Poetry in the CDK and pip in the connector, at least for a transition period

**Review by flash1293 — COMMENTED** [review]
Thanks for starting this, I'm a huge fan of lockfiles to ensure reproducible builds.

I guess in a second phase we would roll out poetry to all python-based connectors?

Left a question, but looks good to me overall (didn't run yet)

File: airbyte-cdk/python/pyproject.toml (hunk 1 above)
Comment (aaronsteers) [reply airbyte-cdk/python/pyproject.toml:63]: Good catch. I experimented with both formats, and missed getting them all consistent.

I believe both formats are (basically?) identical - either declaring the name of the extra inline with the  dependency or declaring the group of dependencies that comprise the extra below as an array of dependent names.

I'll fix so they all use the same convention.

File: airbyte-cdk/python/README.md (hunk 2 above)
Comment (aaronsteers) [reply airbyte-cdk/python/README.md:112]: Yep! `pip install -e .` should still work with Poetry packages. 👍

File: airbyte-cdk/python/setup.py
Comment (bnchrch) [inline airbyte-cdk/python/setup.py:1]: Always so happy to see this gone

File: airbyte-cdk/python/pyproject.toml [hunk 3]
```diff
@@ -1,8 +1,91 @@
...
+    "airbyte_cdk",
+    "airbyte_cdk/py.typed",
+    "airbyte_cdk/sources/declarative/declarative_component_schema.yaml"
+]
+
+[tool.poetry.urls]
+"Tracker" = "https://github.com/airbytehq/airbyte/issues"
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<4.0"
```
Comment (bnchrch) [inline airbyte-cdk/python/pyproject.toml:30]: Should be able to use `^` here instead

```suggestion
python = "^3.8.1
```

**Review by bnchrch — APPROVED** [review]
Great to see!

**Comment by aaronsteers** [comment]
Thanks to everyone who has reviewed so far. 

I have a PR based on this one where I'm testing the interface with `airbyte-ci` and the ability to build when a connector (also Poetry in this case) needs to be built using `--use-local-cdk`. So far, it is not quite working but I'm trying a few different things.

**Comment by natikgadzhi** [comment]
Oh wow. We migrated the whole thing to Poetry and haven't even stumbled at this PR 🤦🏼. Closing since we already got this done.

**Pull request closed without merging**